common/flatpak-dir.c

```diff
@@ -1010,66 +1010,273 @@ flatpak_ensure_oci_summary_cache_dir_location (GError **error)
   return g_steal_pointer (&dir);
 }
 
-static FlatpakSystemHelper *
-flatpak_dir_get_system_helper (FlatpakDir *self)
+static gboolean
+flatpak_dir_use_system_helper (FlatpakDir *self,
+                               const char *installing_from_remote)
 {
-  g_autoptr(GError) error = NULL;
+#ifndef USE_SYSTEM_HELPER
+  if (TRUE)
+    return FALSE;
+#endif
+
+  if (self->no_system_helper || self->user || getuid () == 0)
+    return FALSE;
+
+  /* OCI doesn't do signatures atm, so we can't use the system helper for this */
+  if (installing_from_remote != NULL && flatpak_dir_get_remote_oci (self, installing_from_remote))
+    return FALSE;
 
-  if (g_once_init_enter (&self->system_helper))
+  return TRUE;
+}
+
+static GVariant *
+flatpak_dir_system_helper_call (FlatpakDir   *self,
+                                const gchar  *method_name,
+                                GVariant     *parameters,
+                                GCancellable *cancellable,
+                                GError      **error)
+{
+  if (g_once_init_enter (&self->system_helper_bus))
     {
-      FlatpakSystemHelper *system_helper;
       const char *on_session = g_getenv ("FLATPAK_SYSTEM_HELPER_ON_SESSION");
+      GDBusConnection *system_helper_bus =
+        g_bus_get_sync (on_session != NULL ? G_BUS_TYPE_SESSION : G_BUS_TYPE_SYSTEM,
+                        cancellable, error);
 
       /* To ensure reverse mapping */
       flatpak_error_quark ();
 
-      system_helper =
-        flatpak_system_helper_proxy_new_for_bus_sync (on_session != NULL ? G_BUS_TYPE_SESSION : G_BUS_TYPE_SYSTEM,
-                                                      G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES |
-                                                      G_DBUS_PROXY_FLAGS_DO_NOT_CONNECT_SIGNALS,
-                                                      "org.freedesktop.Flatpak.SystemHelper",
-                                                      "/org/freedesktop/Flatpak/SystemHelper",
-                                                      NULL, &error);
-      if (error != NULL)
-        {
-          g_warning ("Can't find org.freedesktop.Flatpak.SystemHelper: %s", error->message);
-          system_helper = NO_SYSTEM_HELPER;
-        }
-      else
-        {
-          g_dbus_proxy_set_default_timeout (G_DBUS_PROXY (system_helper), G_MAXINT);
-        }
-      g_once_init_leave (&self->system_helper, system_helper);
+      g_once_init_leave (&self->system_helper_bus, system_helper_bus);
```

Comment:
`g_once_init_leave()` is documented to take a nonzero result. Make `system_helper_bus` a static variable and set a separate `static gsize` to 1 for the execute-once part?

Reply:
I don’t think that is right. It will not get the memory barriers right. I could set system_helper_bus to 1 in the failure case though, arch check for that as failure.

Reply:
Sorry, I wanted to give a quick negative review for the incorrect use of `g_once_init_leave()`, but it was too late at night to give a correct solution. What I was trying to suggest was separating the variable used for the once-init from the variable for the bus, something like this:

```
static gsize once = 0;

if (g_once_init_enter (&once)
  {
    self->system_helper_bus = ... (&self, system_helper_bus->error);
    g_once_init_leave (&once, 1);
  }

if (self->system_helper_bus == NULL)
  /* error */
```

But `g_bus_get_sync()` is idempotent so maybe this is premature optimization anyway.